p4include/psa.p4

```diff
@@ -0,0 +1,672 @@
+ * have an instance for each entry in the table.
+ */
+
+/// Counter types
+enum CounterType_t {
+    packets,
+    bytes,
+    packets_and_bytes
+}
+
+/// Counter: RFC-XXXX
+extern Counter<W, S> {
+  Counter(S n_counters, W size_in_bits, CounterType_t counter_type);
+  void count(in S index, in W increment);
+
+  /*
+  @ControlPlaneAPI
+  {
+    W    read<W>      (in S index);
+    W    sync_read<W> (in S index);
+    void set          (in S index, in W seed);
+    void reset        (in S index);
+    void start        (in S index);
+    void stop         (in S index);
+  }
+  */
+}
+
+/// DirectCounter: RFC-XXXX
+extern DirectCounter<W> {
+  DirectCounter(entry_key key, CounterType_t counter_type);
+  void count();
+
+  /*
+  @ControlPlaneAPI
+  {
+    W    read<W>      (in entry_key key);
+    W    sync_read<W> (in entry_key key);
+    void set          (in W seed);
+    void reset        (in entry_key key);
+    void start        (in entry_key key);
+    void stop         (in entry_key key);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Meters: RFC 2698
+ *
+ * Meters are a more complex mechanism for keeping statistics about
+ * the packets that trigger a table. The meters specified in the PSA
+ * are 3-color meters.
+ */
+
+/// Meter types
+enum MeterType_t {
+    packets,
+    bytes
+}
+/// Meter colors
+enum MeterColor_t { RED, GREEN, YELLOW };
+
+/// Meter
+extern Meter<S> {
+  Meter(S n_meters, MeterType_t type);
+  MeterColor_t execute(in S index, in MeterColor_t color);
+
+  /*
+  @ControlPlaneAPI
+  {
+    reset(in MeterColor_t color);
+    setParams(in S committedRate, in S committedBurstSize
+              in S peakRate, in S peakBurstSize);
+    getParams(out S committedRate, out S committedBurstSize
+              out S peakRate, out S peakBurstSize);
+  }
+  */
+}
+
+/// DirectMeter
+extern DirectMeter {
+  DirectMeter(MeterType_t type);
+  MeterColor_t execute(in MeterColor_t color);
+
+  /*
+  @ControlPlaneAPI
+  {
+    reset(in MeterColor_t color);
+    void setParams<S>(in S committedRate, in S committedBurstSize
+                      in S peakRate, in S peakBurstSize);
+    void getParams<S>(out S committedRate, out S committedBurstSize
+                      out S peakRate, out S peakBurstSize);
+  }
+  */
+}
+
+/* --------------------------------------------------------------------- */
+/**
+ * Registers
+ *
+ * Registers are stateful memories whose values can be read and
+ * written in actions. Registers are similar to counters, but can be
+ * used in a more general way to keep state.
+ *
+ * Although registers cannot be used directly in matching,
+ * register.read may be used as the RHS of an assignment operation,
+ * allowing the current value of the register to be copied into
+ * metadata and be available for matching in subsequent tables.
+ */
+extern Register<T, S> {
+  Register(S size);
```

Comment:
I wonder if the constructor should be 
```
    Register<T>(S instance_count);
```
instead.

The reason is that the type of the value stored in the register should be provided to the constructor, not derived from usage.

Also, as a minor comment, I think that ```(S size)``` is not a very good name, because it is not clear whether this refers to the size of an individual register or the size of the register *array*. I propose we use ```(I instance_count)``` in the declaration and ```(I index)``` in the methods.